components/places/src/import/fennec_bookmarks.rs

```diff
@@ -0,0 +1,300 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::api::places_api::PlacesApi;
+use crate::bookmark_sync::{
+    store::{BookmarksStore, Merger},
+    SyncedBookmarkKind,
+};
+use crate::error::*;
+use crate::import::common::{attached_database, ExecuteOnDrop};
+use crate::types::SyncStatus;
+use url::Url;
+
+pub fn import_fennec_bookmarks(
+    places_api: &PlacesApi,
+    path: impl AsRef<std::path::Path>,
+) -> Result<()> {
+    let url = crate::util::ensure_url_path(path)?;
+    do_import_fennec_bookmarks(places_api, url)
+}
+
+fn do_import_fennec_bookmarks(places_api: &PlacesApi, fennec_db_file_url: Url) -> Result<()> {
+    let conn = places_api.open_sync_connection()?;
+
+    let scope = conn.begin_interrupt_scope();
+
+    sql_fns::define_functions(&conn)?;
+
+    // Not sure why, but apparently beginning a transaction sometimes
+    // fails if we open the DB as read-only. Hopefully we don't
+    // unintentionally write to it anywhere...
+    // fennec_db_file_url.query_pairs_mut().append_pair("mode", "ro");
+
+    log::trace!("Attaching database {}", fennec_db_file_url);
+    let auto_detach = attached_database(&conn, &fennec_db_file_url, "fennec")?;
+
+    let tx = conn.begin_transaction()?;
+
+    let clear_mirror_on_drop = ExecuteOnDrop::new(&conn, WIPE_MIRROR.to_string());
+
+    // Clear the mirror now, since we're about to fill it with data from the fennec
+    // connection.
+    log::debug!("Clearing mirror to prepare for import");
+    conn.execute_batch(&WIPE_MIRROR)?;
+    scope.err_if_interrupted()?;
+
+    log::debug!("Populating mirror with the bookmarks roots");
+    crate::bookmark_sync::create_synced_bookmark_roots(&conn)?;
+    scope.err_if_interrupted()?;
+
+    log::debug!("Creating staging table");
+    conn.execute_batch(&CREATE_STAGING_TABLE)?;
```

Comment:
Hmm, I'm curious why we need the staging table. Can we insert directly from `fennec.bookmarks` into `moz_bookmarks_synced` and `moz_bookmarks_synced_structure`? We'd need to do two passes, one to insert items into `moz_bookmarks_synced` and the other to insert structure into `moz_bookmarks_synced_structure`, but, IIUC, we already need to do two passes over the staging table, anyway?

We'd need to validate and insert `fennec.bookmarks.url` into `moz_places` first, then associate those URLs with bookmark items. (For invalid URLs, we can set the `validity` column in `moz_bookmarks_synced` to `Replace`, and Dogear will take care of deleting it).